### PR TITLE
PTL framework accepts special characters in job_name string

### PR DESCRIPTION
**Pull request by birbalsain**
#### Describe Bug or Feature
When we submit job with job name like "qsub -N jobname_string" and jobname_string has special character but we want special character as a ordinary character then cmdline unable to read special character as a ordinary character.

#### Describe Your Change
Prefixing a special character with "\\" to turn it into ordinary character.

#### Attach Test and Valgrind Logs/Output

[qsub_N_cmdline_platform_1.txt](https://github.com/PBSPro/pbspro/files/4179262/qsub_N_cmdline_platform_1.txt)
[qsub_N_cmdline_platform_2.txt](https://github.com/PBSPro/pbspro/files/4179263/qsub_N_cmdline_platform_2.txt)

### CODE REVIEW

**Comment by birbalsain** [comment]
@PBSPro/pbspro-maintainers please take this PR further.

**Comment by birbalsain** [comment]
@agrawalravi90 test is not running on some platform. I attached logs of two different platforms. 

May be you are not using the platform where test was failed.

**Comment by agrawalravi90** [comment]
Ok, but with your changes, the -N string now has additional '\' characters, NOT escaped '&'s:
**2020-02-04 19:34:15,176 INFOCLI  pbspro: sudo -H -u pbsuser /opt/pbs/bin/qsub -N j\&whoami\&gt\;/tmp/b\& -- /bin/sleep 100
2020-02-04 19:34:15,201 INFO     submit to pbspro as pbsuser: job OrderedDict([('Job_Name', 'j\\&whoami\\&gt\\;/tmp/b\\&')])**

Especially by making them 'raw' strings, you are ensuring that the slashes will not escape special characters.

if this is needed for a special platform like shasta or cray, then instead of changing the tests, it might be better to handle this in the framework as this will useful for other tests passing in special characters in strings as well. Search for "self.platform == 'shasta'" in our codebase to see examples of how we handle other platform specific stuff in ptl.

**Comment by bhroam** [comment]
@birbalsain Could you make your PR description a little more descriptive?  I was unable to figure out what you were doing from the oneline you gave.

**Comment by birbalsain** [comment]
@agrawalravi90 I have done the requested changes.

File: test/fw/ptl/lib/pbs_testlib.py [hunk 1]
```diff
@@ -2383,6 +2383,9 @@ def convert_to_cli(self, attrs, op=None, hostname=None, dflt_conf=True,
             attrs = attrs.items()
 
         for a, v in attrs:
+            if a == "Job_Name":
```
Comment (agrawalravi90) [inline test/fw/ptl/lib/pbs_testlib.py]: This adds an extra slash before every special character, not escape it, for centos8:
`sudo -H -u pbsuser /opt/pbs/bin/qsub -N j\&whoami\&gt\;\/tmp\/b\& -- /bin/sleep 100`
So, i think you need to make this platform specific for only the platforms which have this issue

File: test/fw/ptl/lib/pbs_testlib.py [hunk 2]
```diff
@@ -2383,6 +2383,9 @@ def convert_to_cli(self, attrs, op=None, hostname=None, dflt_conf=True,
...
+                v = v.translate({ord(c): "\\" +
```
Comment (agrawalravi90) [inline test/fw/ptl/lib/pbs_testlib.py]: please add a comment to explain what this line is doing, it's not obvious.

File: test/tests/functional/pbs_validate_job_qsub_attributes.py [hunk 3]
```diff
@@ -133,7 +133,11 @@ def test_qsub_N_job_array(self):
         This test case validates, illlegal characters in job name
         for a job array, cause qsub to error out
         """
-        J = Job(TEST_USER, attrs={ATTR_N: 'j&whoami&g;/tmp/b&', ATTR_J: '1-2'})
+        J = Job(
```
Comment (agrawalravi90) [inline test/tests/functional/pbs_validate_job_qsub_attributes.py]: why did you change the formatting of this line? It looks much harder to read now. I suggest reverting it to how it was before.

**Comment by birbalsain** [comment]
@agrawalravi90 i have done the changes.

**Comment by agrawalravi90** [comment]
@birbalsain please post latest test logs from the relevant platforms

**Comment by birbalsain** [comment]
@agrawalravi90 I have updated latest test logs.

**Comment by agrawalravi90** [comment]
@birbalsain thanks, please get one more sign-off and then we can merge this

File: test/fw/ptl/lib/pbs_testlib.py [hunk 4]
```diff
@@ -2383,6 +2384,14 @@ def convert_to_cli(self, attrs, op=None, hostname=None, dflt_conf=True,
             attrs = attrs.items()
 
         for a, v in attrs:
+            # In job name string, use prefix "\" with special charater
+            # to read as an ordinary character on
+            # cray, carysim and shasta platform
```
Comment (mike0042) [inline test/fw/ptl/lib/pbs_testlib.py]: s/carysim/craysim/

Comment (birbalsain) [reply test/fw/ptl/lib/pbs_testlib.py]: @mike0042 your mean i should use "or" instead of "and" or something else. because i didn't get "carysim" ans "s"

Comment (mike0042) [reply test/fw/ptl/lib/pbs_testlib.py]: No, it's vim syntax. Please change "carysim" to "craysim". And please add a comma after "craysim" once it's changed. I favor use of the Oxford comma when writing lists.

**Comment by birbalsain** [comment]
Thank you @mike0042

**Comment by birbalsain** [comment]
@PBSPro/pbspro-maintainers Please take this PR further.
